winrm/x509.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2016 Canonical Ltd.
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package winrm
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/utils"
+	"github.com/juju/utils/cert"
+)
+
+// List of exported internal errors
+var (
+	ErrNoClientCert       = errors.New("No client winrm cert inside dir")
+	ErrNoClientPrivateKey = errors.New("No client private key inside dir")
+	ErrNoX509Folder       = errors.New("No configuration x509 folder is present in this location")
+)
+
+// x509WinRM type that defines windows remote manager
+// credentials for client-server secure communication
+type x509 struct {
+	cert   []byte // client cert
+	key    []byte // client private key
+	cacert []byte // ca server cert
+	err    error  // internal err state
+}
+
+var (
+	mu          sync.Mutex
+	credentials x509 //winrm credentials
+)
+
+// LoadClientCert generates client cert for x509 authentication
+// if the directory in the coresponding paths key, cert
+// if the base dir dosen't exist it will create and generate the key/cert pair
+// LoadClientCert function is lock protected allowing multiple threads to call this function
+func LoadClientCert(key, cert string) error {
+	mu.Lock()
+	defer mu.Unlock()
+	b1, key := filepath.Split(key)
+	b2, cert := filepath.Split(cert)
+	if strings.Compare(b1, b2) != 0 {
+		return fmt.Errorf("Cert and Key base paths dosen't match")
+	}
+
+	base, err := utils.NormalizePath(b1)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Init winrm credentials path for the module %s", base)
+	logger.Debugf("Init winrm path key %s", key)
+	logger.Debugf("Init winrm path cert %s", cert)
+
+	credentials.read(base, key, cert)
+
+	if err := os.MkdirAll(base, 0700); err != nil {
+		return err
+	}
+
+	credentials.write(base, key, cert)
+
+	// after doing all the work if somthing when wrong
+	// remove all config, and when the command will run
+	// again regenerating new client cert, and key
+	defer func() {
+		if credentials.err != nil {
+			logger.Warningf("An error occured while parsing the winrm keys %s", credentials.err)
+			credentials.err = os.RemoveAll(base)
+		}
+	}()
+
+	return credentials.error()
+}
+
+// LoadCACert saves into internal singleton credentials the CACert to be used for https connections
+// if the cacert in the path is not found it will return the error rather than saing it into state
+// This behaviour will encourage that https connection will skip checking the CA.
+func LoadCACert(path string) error {
```

Comment:
As before, please don't use globals. It seems unlikely that we would have a well-defined location for a CA cert on disk, as it's going to be different for each server. In Juju, I guess there's two places where we would need to specify a CA cert:
 - when manual-provisioning, using a CA cert specified by the user on the command line
 - when doing the equivalent of "juju ssh" ("juju winrm"?), using a CA cert published by the Juju agent on the target machine